File: src/lib/duel.ts
import { daysBetween, omanDate, addDays } from './time';

/**
 * Scoring for a head-to-head challenge.
 *
 * Pure and free of Prisma so the rules can be tested directly — the ways a
 * duel goes wrong are all arithmetic (a tie misreported as a win, a day
 * counted twice, a lead change fired on every save) and none of them need a
 * database to reproduce.
 *
 * Two metrics score differently and the difference matters:
 *
 *   - Volume metrics (steps, workouts, water) total everything logged. Beating
 *     the daily target is worth something, so 12,000 steps counts as 12,000.
 *   - STREAK is about not breaking. Only whole days that met the target count,
 *     and going twice as far on Tuesday does not buy Wednesday off.
 */

export type DuelMetric = 'STEPS' | 'WORKOUTS' | 'WATER' | 'STREAK' | 'CUSTOM';

export type Entry = { userId: string; day: number; amount: number };

export type DuelShape = {
  metric: DuelMetric;
  /** Daily goal. For STREAK this is what a day must reach to count. */
  target: number;
  durationDays: number;
  startedAt: Date;
};

/** Whether a day's logged amount satisfies that day's goal. */
export function dayMet(amount: number, target: number): boolean {
  return target > 0 && amount >= target;
}

/**
 * One side's score.
 *
 * Amounts are clamped to the target for STREAK only. For volume metrics the
 * excess is the whole point of competing.
 */
export function scoreFor(
  duel: DuelShape,
  entries: readonly Entry[],
  userId: string,
): number {
  const mine = entries.filter((e) => e.userId === userId && e.day >= 1 && e.day <= duel.durationDays);

  if (duel.metric === 'STREAK') {
    return mine.filter((e) => dayMet(e.amount, duel.target)).length;
  }

  return mine.reduce((sum, e) => sum + Math.max(0, e.amount), 0);
}

/** Which day of the duel it is, 1-based, clamped to the duel's length. */
export function duelDay(duel: DuelShape, now: Date = new Date()): number {
  const raw = daysBetween(duel.startedAt, now) + 1;
  return Math.min(Math.max(1, raw), duel.durationDays);
}

/**
 * The instant the duel closes: Oman midnight at the end of its final day.
 *
 * `omanDate` returns a UTC-midnight marker for a calendar day, which is four
 * hours *later* than the Muscat midnight it stands for. Returning it unadjusted
 * kept a finished duel accepting entries until 4am the next morning — long
 * enough for someone to log a winning number after the contest had ended.
 */
export function duelEndsAt(startedAt: Date, durationDays: number): Date {
  const dayAfterLast = addDays(omanDate(startedAt), durationDays);
  return new Date(dayAfterLast.getTime() - OMAN_OFFSET_MS);
}

const OMAN_OFFSET_MS = 4 * 60 * 60_000;

/**
 * Whether the duel is over.
 *
 * Compared by day number rather than by instant: both sides are already
 * bucketed into Oman calendar days, and mixing the two representations is
 * exactly what produced the four-hour overrun above.
 */
export function hasEnded(duel: DuelShape, now: Date = new Date()): boolean {
  return daysBetween(duel.startedAt, now) + 1 > duel.durationDays;
}

export type Standing = {
  challengerScore: number;
  opponentScore: number;
  /** Positive when the challenger leads, negative when the opponent does. */
  margin: number;
  /** Null while level — a tie has no leader, and pretending otherwise is how
   *  "took the lead" notifications end up firing at nobody. */
  leaderId: string | null;
};

export function standing(
  duel: DuelShape,
  entries: readonly Entry[],
  challengerId: string,
  opponentId: string,
): Standing {
  const challengerScore = scoreFor(duel, entries, challengerId);
  const opponentScore = scoreFor(duel, entries, opponentId);
  const margin = challengerScore - opponentScore;

  return {
    challengerScore,
    opponentScore,
    margin,
    leaderId: margin === 0 ? null : margin > 0 ? challengerId : opponentId,
  };
}

export type Result =
  | { outcome: 'DRAW'; winnerId: null }
  | { outcome: 'WIN'; winnerId: string };

/**
 * The final result. A level score is a draw, explicitly — never resolved by
 * who logged first or who was challenged, both of which are arbitrary and
 * would feel rigged to the person who lost on them.
 */
export function settle(
  duel: DuelShape,
  entries: readonly Entry[],
  challengerId: string,
  opponentId: string,
): Result {
  const { leaderId } = standing(duel, entries, challengerId, opponentId);
  return leaderId === null
    ? { outcome: 'DRAW', winnerId: null }
    : { outcome: 'WIN', winnerId: leaderId };
}

/** Points awarded when a duel settles. */
export const DUEL_POINTS = {
  /** Both sides get this for finishing, win or lose. */
  COMPLETION: 75,
  /** The winner's additional share. A draw pays neither side the bonus. */
  WIN_BONUS: 75,
} as const;

export function pointsFor(result: Result, userId: string): number {
  const won = result.outcome === 'WIN' && result.winnerId === userId;
  return DUEL_POINTS.COMPLETION + (won ? DUEL_POINTS.WIN_BONUS : 0);
}

File: mobile/lib/features/challenges/domain/duel.dart
import '../../../l10n/app_localizations.dart';

/// What a duel measures.
///
/// Volume metrics total everything logged, because beating the daily target is
/// the point of competing. STREAK counts whole days only — going twice as far
/// on Tuesday does not buy Wednesday off, and not breaking is the whole skill
/// being tested.
enum DuelMetric {
  steps(defaultTarget: 10000),
  workouts(defaultTarget: 1),
  water(defaultTarget: 8),
  streak(defaultTarget: 1);

  const DuelMetric({required this.defaultTarget});

  final int defaultTarget;

  bool get countsWholeDaysOnly => this == DuelMetric.streak;

  String title(L l) => switch (this) {
        DuelMetric.steps => l.metricSteps,
        DuelMetric.workouts => l.metricWorkouts,
        DuelMetric.water => l.metricWater,
        DuelMetric.streak => l.metricStreak,
      };

  String subtitle(L l) => switch (this) {
        DuelMetric.steps => l.metricStepsSub,
        DuelMetric.workouts => l.metricWorkoutsSub,
        DuelMetric.water => l.metricWaterSub,
        DuelMetric.streak => l.metricStreakSub,
      };
}

/// A person you can compete with.
///
/// Only what is public. A challenge profile never carries a phone number, an
/// email, a location or anything about health — those exist on the account and
/// have no business on a screen another member can open.
class Opponent {
  const Opponent({
    required this.name,
    required this.username,
    required this.level,
    required this.wins,
    required this.streak,
    this.acceptsChallenges = true,
  });

  final String name;
  final String username;
  final int level;
  final int wins;
  final int streak;

  /// False when they have closed the door. The button is hidden, and a real
  /// server would refuse the request too — a hidden button stops only the
  /// polite path.
  final bool acceptsChallenges;
}

/// One side's running total.
class DuelSide {
  const DuelSide({required this.name, required this.score});
  final String name;
  final int score;
}

/// Where a duel stands.
class DuelStanding {
  const DuelStanding({required this.mine, required this.theirs});

  final DuelSide mine;
  final DuelSide theirs;

  int get margin => mine.score - theirs.score;

  /// Null while level. A tie has no leader, and inventing one is how a
  /// "took the lead" notification ends up firing at nobody.
  DuelSide? get leader => switch (margin) {
        0 => null,
        > 0 => mine,
        _ => theirs,
      };

  bool get iAmLeading => leader == mine;

  /// Share of the combined total, for the split bar. Falls back to level when
  /// neither side has logged anything, rather than dividing by zero.
  double get share {
    final total = mine.score + theirs.score;
    return total == 0 ? 0.5 : mine.score / total;
  }
}
